src/main/java/org/folio/circulation/domain/notice/schedule/DueDateScheduledNoticeHandler.java

```diff
@@ -70,12 +84,46 @@ public DueDateScheduledNoticeHandler(
 
   private CompletableFuture<Result<ScheduledNotice>> handleDueDateNotice(ScheduledNotice notice) {
     return loanRepository.getById(notice.getLoanId())
+      .thenCompose(r -> deleteNoticeIfLoanIsMissingOrIncomplete(r, notice))
       .thenApply(r -> r.map(LoanAndRelatedRecords::new))
       .thenCompose(r -> r.after(loanPolicyRepository::lookupLoanPolicy))
-      .thenCompose(r -> r.combineAfter(configurationRepository::findTimeZoneConfiguration,
-        LoanAndRelatedRecords::withTimeZone))
+      .thenCompose(r -> r.combineAfter(configurationRepository::findTimeZoneConfiguration, LoanAndRelatedRecords::withTimeZone))
       .thenCompose(r -> r.after(records -> sendNotice(records, notice)))
-      .thenCompose(r -> r.after(relatedRecords -> updateNotice(relatedRecords, notice)));
+      .thenCompose(r -> r.after(relatedRecords -> updateNotice(relatedRecords, notice)))
+      .thenApply(r -> r.mapFailure(this::handleFailure));
+  }
+
+  CompletableFuture<Result<Loan>> deleteNoticeIfLoanIsMissingOrIncomplete(
+      Result<Loan> result, ScheduledNotice notice) {
+
+    if (failedToFindRecordOfType(result, LOAN_REC_TYPE)) {
```

Comment:
It seems to me that we will need to add/remove some 'rules' when a notice should be removed.

I it possible to introduce a `shouldDeleteNotice` method. And depending on the result of the method do deletion or not.

Reply:
Here we not only decide whether we should delete a notice, but also determine which referenced entity is missing. And since both of these tasks share the same condition, it makes sense to perform them simultaneously. In this case extracting this new method might not be the best idea.

Reply:
Thanks for explaining. Let's leave it as is.